### PR TITLE
Add Tensor::CopyFrom and Tensor::mutable_data(Place place)

### PR DESCRIPTION
**Pull request by Canpio**
1. Add `Tensor::CopyFrom`. Current version can only support CPU memory
copy. The support of GPU will be provided later by `paddle::memory`.
The current implementation of `Tensor::CopyFrom` is a little inefficient:
Every time `CopyFrom` is called, tensor will re-allocate its memory. However, if
we try to check and reuse `placeholder_`, we have to provide a template
parameter for `CopyFrom` to indicate the data type. It seems strange for
a simple copy function.

2. Add `Tensor::mutable_data(Place place)`, which directly use member
variable `dims_` as its dim parameter. This interface is required by
`Op::InferShape`.

### CODE REVIEW

File: paddle/framework/tensor_test.cc [hunk 1]
```diff
@@ -178,4 +178,29 @@ TEST(Tensor, Slice) {
...
+  Tensor dst_tensor;
+  dst_tensor.CopyFrom(src_tensor, CPUPlace());
+  const int* dst_ptr = dst_tensor.data<int>();
+  ASSERT_NE(src_ptr, dst_ptr);
+  for (size_t i = 0; i < 9; ++i) {
+    EXPECT_EQ(src_ptr[i], dst_ptr[i]);
+  }
+
+  Tensor slice_tensor = src_tensor.Slice(1, 2);
+  dst_tensor.CopyFrom(slice_tensor, CPUPlace());
```
Comment (luotao1) [inline paddle/framework/tensor_test.cc]: 这个CopyFrom和memcpy还是不大一样吧。比如`memcpy(src_ptr, att + 1, 8 *sizeof(int))`，CopyFrom能做么？是准备后续增加？

Comment (Canpio) [inline paddle/framework/tensor_test.cc]: 我的理解是，这种可以随意指定开头和长度的拷贝对于tensor来说并不一定有意义，因为很可能拷贝出来的内容无法组成一个tensor，所以并不应该是CopyFrom应该提供的功能

Comment (luotao1) [inline paddle/framework/tensor_test.cc]: Got it. 同意

File: paddle/framework/tensor.h [hunk 2]
```diff
@@ -44,11 +45,17 @@ class Tensor {
...
+  template <typename T,  // must be POD types
+            typename std::enable_if<std::is_pod<T>::value>::type* = nullptr>
+  T* mutable_data(paddle::platform::Place place) {
     if (holder_ == nullptr ||
         !(holder_->Place() ==
           place) /* some versions of boost::variant don't have operator!= */
-        || holder_->Size() < product(dims) * sizeof(T) + offset_) {
-      holder_.reset(new PlaceholderImpl<T>(place, product(dims) * sizeof(T)));
+        || holder_->Size() < product(dims_) * sizeof(T) + offset_) {
+      holder_.reset(new PlaceholderImpl<T>(place, product(dims_) * sizeof(T)));
```
Comment (qingqing01) [inline paddle/framework/tensor.h]: > Tensor() : offset_(0) {}

这个构造函数没有初始化dims_，假如先调用这个空的构造函数，再调用该mutable_data会有问题。

File: paddle/framework/tensor.h [hunk 3]
```diff
@@ -63,6 +70,15 @@ class Tensor {
     offset_ = src.offset_;
   }
 
+  void CopyFrom(const Tensor& src, paddle::platform::Place dst_place) {
```
Comment (qingqing01) [inline paddle/framework/tensor.h]: `dst_place`和`place_`可能不一样？ 为什么要加dst_place？

File: paddle/framework/tensor.h [hunk 4]
```diff
@@ -63,6 +70,15 @@ class Tensor {
...
+    PADDLE_ENFORCE(src.holder_ != nullptr,
+                   "Can not copy from an uninitialized tensor.");
+    size_t size = product(src.dims()) * src.holder_->TypeSize();
```
Comment (qingqing01) [inline paddle/framework/tensor.h]: 很多地方用了product()，觉得加一个获取size的接口也可以~

Comment (qingqing01) [inline paddle/framework/tensor.h]: `* src.holder_->TypeSize()` 也可以在PlaceholderImpl里吧

File: paddle/framework/tensor.h (hunk 2 above)
Comment (Canpio) [inline paddle/framework/tensor.h]: 确实会有这样的问题……看来需要在`mutable_data(Place place)`中`PADDLE_ENFORCE`一下：
```cpp
PADDLE_ENFORCE(product(dims_) > 0)
```

即使不提供`mutable_data(Place place)`接口，这个判断应该也是有必要的，因为在调用`mutable_data(DDim dims, Place place)`的时候也可能传一个`{0}`的DDim进来。

File: paddle/framework/tensor.h (hunk 3 above)
Comment (Canpio) [inline paddle/framework/tensor.h]: 确实可能不一样，可能跨设备拷贝。

File: paddle/framework/tensor.h (hunk 4 above)
Comment (Canpio) [inline paddle/framework/tensor.h]: 确实。另外我怀疑product可能存在性能问题，因为目前的实现是把`DDim`转化为`vector<int>`，再做连乘，可能会比较慢

Comment (Canpio) [inline paddle/framework/tensor.h]: 这个没太理解？

Comment (luotao1) [inline paddle/framework/tensor.h]: 同意获取size的接口。另外，能加个reshape dim的接口么，比如把dim<1, 10, 20>变成<10, 20>。

File: paddle/framework/tensor.h [hunk 5]
```diff
@@ -78,23 +96,32 @@ class Tensor {
...
     return dst;
   }
 
+  void set_dims(const DDim& dims) {
+    if (dims == dims_) {
+      return;
+    }
+    dims_ = dims;
+    numel_ = product(dims_);
+    return;
```
Comment (QiJune) [inline paddle/framework/tensor.h]: return here is unnecessary here, pPlease remove it

File: paddle/framework/tensor.h [hunk 6]
```diff
@@ -118,18 +145,27 @@ class Tensor {
...
     std::unique_ptr<T, Deleter> ptr_;
     paddle::platform::Place place_;  // record the place of ptr_.
     size_t size_;                    // size of the memory block.
   };
 
+  template <typename T>
+  inline void CheckDimsValidity() const {
+    PADDLE_ENFORCE(holder_ != nullptr,
+                   "Tenosr holds no memory. Call Tensor::mutable_data first.");
+    PADDLE_ENFORCE(holder_->size() > numel_ * sizeof(T) + offset_,
```
Comment (QiJune) [inline paddle/framework/tensor.h]: ```>=``` will be good. 
And maybe CheckDims is more concise than CheckDimsValidity

File: paddle/framework/tensor.h [hunk 7]
```diff
@@ -27,45 +27,63 @@ namespace framework {
...
 
+  template <typename T>
+  void CopyFrom(const Tensor& src, paddle::platform::Place dst_place) {
+    PADDLE_ENFORCE(platform::is_cpu_place(src.holder_->place()) &&
+                       platform::is_cpu_place(dst_place),
+                   "Tensor::CopyFrom only support CPU now.");
+    src.CheckDimsValidity<T>();
+    size_t size = src.numel_ * sizeof(T);
+    set_dims(src.dims());
+    const void* src_ptr = static_cast<const void*>(src.data<T>());
```
Comment (QiJune) [inline paddle/framework/tensor.h]: why use ```const``` here(static_cast<const void*>)? data<T>() will return a ```const``` T*

Comment (Canpio) [inline paddle/framework/tensor.h]: `const T*` is not allowed to transform to `void*` by static_cast.

File: paddle/framework/tensor.h (hunk 5 above)
Comment (Canpio) [inline paddle/framework/tensor.h]: done

File: paddle/framework/tensor.h (hunk 6 above)
Comment (Canpio) [inline paddle/framework/tensor.h]: done

File: paddle/framework/tensor.h [hunk 8]
```diff
@@ -27,45 +27,63 @@ namespace framework {
...
-    dims_ = dims;
+    set_dims(dims);
+    return mutable_data<T>(place);
+  }
+
+  template <typename T>
+  T* mutable_data(paddle::platform::Place place) {
+    PADDLE_ENFORCE(numel_ > 0,
+                   "Tensor::numel_ must be larger than zero to call "
+                   "Tensor::mutable_data.");
```
Comment (QiJune) [inline paddle/framework/tensor.h]: Users can not call this method, only he calls set_dims or mutable_data(dim,place) first. So, we should  inform the users more clearly.

And is mutable_data(dim, place) necessary? It's just the combination of set_dims and mutable_data(place).

Comment (Canpio) [inline paddle/framework/tensor.h]: done

**Pull request merged**